{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ac09bf60",
   "metadata": {},
   "source": [
    "# Notebook for MuData Creation from GE and ATSE AnnData\n",
    "\n",
    "This notebook:\n",
    "1. Reads and inspects ATSE and gene expression AnnData files.\n",
    "2. Fixes NaNs in the splicing data.\n",
    "3. Creates modality-specific `.obs`, `.var`, and `.layers` for each AnnData.\n",
    "4. Creates a MuData object with modalities “rna”, “junc_counts”, “cell_by_junction_matrix”, \n",
    "    and “cell_by_cluster_matrix”.\n",
    "5. Writes out the final MuData object for use with MULTIVISPLICE."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "17235178",
   "metadata": {},
   "source": [
    "## 0. Set Paths and Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2c90aa21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATSE data path: /gpfs/commons/groups/knowles_lab/Karin/Leaflet-analysis-WD/MOUSE_SPLICING_FOUNDATION/MODEL_INPUT/052025/aligned_splicing_data_20250513_035938.h5ad\n",
      "GE data path:   /gpfs/commons/groups/knowles_lab/Karin/Leaflet-analysis-WD/MOUSE_SPLICING_FOUNDATION/MODEL_INPUT/052025/aligned_gene_expression_data_20250513_035938.h5ad\n",
      "Output MuData path: /gpfs/commons/groups/knowles_lab/Karin/Leaflet-analysis-WD/MOUSE_SPLICING_FOUNDATION/MODEL_INPUT/052025/SUBSETTOP5CELLSTYPES_aligned__ge_splice_combined_20250513_035938.h5mu\n"
     ]
    }
   ],
   "source": [
    "ROOT_PATH = \"/gpfs/commons/groups/knowles_lab/Karin/Leaflet-analysis-WD/MOUSE_SPLICING_FOUNDATION/MODEL_INPUT/052025/\"\n",
    "\n",
    "ATSE_DATA_PATH = ROOT_PATH + \"aligned_splicing_data_20250513_035938.h5ad\"\n",
    "GE_DATA_PATH = ROOT_PATH + \"aligned_gene_expression_data_20250513_035938.h5ad\"\n",
    "OUTPUT_MUDATA_PATH = ROOT_PATH + \"SUBSETTOP5CELLSTYPES_aligned__ge_splice_combined_20250513_035938.h5mu\"\n",
    "REDO_JUNC_RATIO = False\n",
    "\n",
    "print(\"ATSE data path:\", ATSE_DATA_PATH)\n",
    "print(\"GE data path:  \", GE_DATA_PATH)\n",
    "print(\"Output MuData path:\", OUTPUT_MUDATA_PATH)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3e0ff02a",
   "metadata": {},
   "source": [
    "## 1. Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cdcfda21-0265-482d-91ed-4c8d75744be0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scipy.sparse as sp\n",
    "import numpy as np\n",
    "from scipy.sparse import csr_matrix, hstack, vstack\n",
    "import h5py\n",
    "import anndata as ad\n",
    "import mudata as mu"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "779463e3",
   "metadata": {},
   "source": [
    "## 2. Load ATSE and Gene Expression AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f39a3b9-b0e9-401a-9430-c1da41a431f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "atse_anndata = ad.read_h5ad(ATSE_DATA_PATH)\n",
    "print(\"ATSE AnnData:\", atse_anndata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a8d69e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GE AnnData: AnnData object with n_obs × n_vars = 157418 × 19022\n",
      "    obs: 'cell_id', 'age', 'cell_ontology_class', 'mouse.id', 'sex', 'tissue', 'dataset', 'batch', 'subtissue_clean', 'broad_cell_type', 'cell_id_index', 'cell_name', 'library_size'\n",
      "    var: 'index', 'gene_name', 'gene_id', 'mean_transcript_length', 'mean_intron_length', 'num_transcripts', 'transcript_biotypes'\n",
      "    obsm: 'X_library_size'\n",
      "    layers: 'length_norm', 'log_norm', 'predicted_log_norm_tms', 'raw_counts'\n",
      "<Compressed Sparse Row sparse matrix of dtype 'float64'\n",
      "\twith 593435257 stored elements and shape (157418, 19022)>\n",
      "  Coords\tValues\n",
      "  (0, 21)\t0.029308323563892142\n",
      "  (0, 54)\t0.0165346193592835\n",
      "  (0, 137)\t0.030781944014048136\n",
      "  (0, 185)\t0.004901960784313725\n",
      "  (0, 340)\t0.0016926201760324984\n",
      "  (0, 372)\t0.01990570979570456\n",
      "  (0, 415)\t0.000690766751093714\n",
      "  (0, 419)\t0.06815537745604965\n",
      "  (0, 455)\t0.0024908538958511714\n",
      "  (0, 482)\t0.10429304713019132\n",
      "  (0, 510)\t0.2241253644314869\n",
      "  (0, 535)\t0.0007496251874062968\n",
      "  (0, 557)\t0.0031923383878691143\n",
      "  (0, 708)\t0.06768808598216328\n",
      "  (0, 804)\t0.07531992687385741\n",
      "  (0, 900)\t0.026785714285714284\n",
      "  (0, 1018)\t0.011564625850340135\n",
      "  (0, 1121)\t0.02127659574468085\n",
      "  (0, 1146)\t0.0007147962830593281\n",
      "  (0, 1147)\t0.009900990099009901\n",
      "  (0, 1178)\t0.007010834926704908\n",
      "  (0, 1180)\t0.061760840998685944\n",
      "  (0, 1189)\t0.0010051934997487018\n",
      "  (0, 1219)\t0.004346776141028737\n",
      "  (0, 1231)\t0.0003624173235480656\n",
      "  :\t:\n",
      "  (157417, 18963)\t0.00841964584029402\n",
      "  (157417, 18965)\t0.02587630757937236\n",
      "  (157417, 18968)\t0.0011985617259288853\n",
      "  (157417, 18969)\t0.03779848806047758\n",
      "  (157417, 18971)\t0.031075201988812924\n",
      "  (157417, 18972)\t0.029427501337613696\n",
      "  (157417, 18974)\t0.017670682730923697\n",
      "  (157417, 18976)\t0.0011268028846153845\n",
      "  (157417, 18987)\t0.0029251984956122023\n",
      "  (157417, 18988)\t0.030022465790727752\n",
      "  (157417, 18990)\t0.029558617934891153\n",
      "  (157417, 18992)\t0.0006774302811335666\n",
      "  (157417, 18996)\t0.004804026231508312\n",
      "  (157417, 18997)\t0.0054575986565911\n",
      "  (157417, 18998)\t0.0006757669955399379\n",
      "  (157417, 19006)\t0.020247469066366708\n",
      "  (157417, 19008)\t0.011816838995568686\n",
      "  (157417, 19010)\t0.0005374417771408097\n",
      "  (157417, 19012)\t0.00546448087431694\n",
      "  (157417, 19013)\t0.003415744957709824\n",
      "  (157417, 19015)\t0.46540880503144655\n",
      "  (157417, 19018)\t0.014664040271394178\n",
      "  (157417, 19019)\t0.01149566607241655\n",
      "  (157417, 19020)\t0.0008469372631222343\n",
      "  (157417, 19021)\t0.0012823252831801666\n"
     ]
    }
   ],
   "source": [
    "ge_anndata = ad.read_h5ad(GE_DATA_PATH)\n",
    "print(\"GE AnnData:\", ge_anndata)\n",
    "print(ge_anndata.layers[\"length_norm\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5579f27e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Compressed Sparse Row sparse matrix of dtype 'float64'\n",
      "\twith 593435257 stored elements and shape (157418, 19022)>\n",
      "  Coords\tValues\n",
      "  (0, 21)\t51.0\n",
      "  (0, 54)\t29.0\n",
      "  (0, 137)\t54.0\n",
      "  (0, 185)\t8.0\n",
      "  (0, 340)\t3.0\n",
      "  (0, 372)\t35.0\n",
      "  (0, 415)\t1.0\n",
      "  (0, 419)\t120.0\n",
      "  (0, 455)\t4.0\n",
      "  (0, 482)\t184.0\n",
      "  (0, 510)\t397.0\n",
      "  (0, 535)\t1.0\n",
      "  (0, 557)\t5.0\n",
      "  (0, 708)\t120.0\n",
      "  (0, 804)\t133.0\n",
      "  (0, 900)\t47.0\n",
      "  (0, 1018)\t20.0\n",
      "  (0, 1121)\t37.0\n",
      "  (0, 1146)\t1.0\n",
      "  (0, 1147)\t17.0\n",
      "  (0, 1178)\t12.0\n",
      "  (0, 1180)\t109.0\n",
      "  (0, 1189)\t1.0\n",
      "  (0, 1219)\t7.0\n",
      "  (0, 1231)\t0.0\n",
      "  :\t:\n",
      "  (157417, 18963)\t14.0\n",
      "  (157417, 18965)\t45.0\n",
      "  (157417, 18968)\t2.0\n",
      "  (157417, 18969)\t67.0\n",
      "  (157417, 18971)\t55.0\n",
      "  (157417, 18972)\t52.0\n",
      "  (157417, 18974)\t31.0\n",
      "  (157417, 18976)\t1.0\n",
      "  (157417, 18987)\t5.0\n",
      "  (157417, 18988)\t53.0\n",
      "  (157417, 18990)\t52.0\n",
      "  (157417, 18992)\t1.0\n",
      "  (157417, 18996)\t8.0\n",
      "  (157417, 18997)\t9.0\n",
      "  (157417, 18998)\t1.0\n",
      "  (157417, 19006)\t35.0\n",
      "  (157417, 19008)\t20.0\n",
      "  (157417, 19010)\t0.0\n",
      "  (157417, 19012)\t9.0\n",
      "  (157417, 19013)\t6.0\n",
      "  (157417, 19015)\t825.0\n",
      "  (157417, 19018)\t25.0\n",
      "  (157417, 19019)\t20.0\n",
      "  (157417, 19020)\t1.0\n",
      "  (157417, 19021)\t2.0\n"
     ]
    }
   ],
   "source": [
    "# rescale by overall median transcript length (didn't do this in preprocessing of GE AnnData)\n",
    "ge_anndata.layers[\"length_norm\"] = ge_anndata.layers[\"length_norm\"] * np.median(ge_anndata.var[\"mean_transcript_length\"])\n",
    "# make sure to round down to get integer counts (this is CSR)\n",
    "ge_anndata.layers[\"length_norm\"].data = np.floor(ge_anndata.layers[\"length_norm\"].data)\n",
    "print(ge_anndata.layers[\"length_norm\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3ad9d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 104530.]\n",
      " [ 415320.]\n",
      " [ 330365.]\n",
      " ...\n",
      " [ 873490.]\n",
      " [1368928.]\n",
      " [ 431221.]]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/home/svaidyanathan/miniconda3/envs/scvi-env/lib/python3.12/site-packages/anndata/_core/storage.py:39: ImplicitModificationWarning: Obsm 'X_library_size' should not be a np.matrix, use np.ndarray instead.\n",
      "  warnings.warn(msg, ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "# Recalculate library size using length normalized counts\n",
    "ge_anndata.obsm[\"X_library_size\"] = ge_anndata.layers[\"length_norm\"].sum(axis=1)\n",
    "print(ge_anndata.obsm[\"X_library_size\"])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c0847406-01c0-4cd5-a976-3bdbe05b56f4",
   "metadata": {},
   "source": [
    "## 3. Create `.var` DataFrames for Each Modality\n",
    "\n",
    "Here we create modality-specific `.var` metadata. You might later use these to update the\n",
    "corresponding AnnData objects inside the MuData container."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07ad3c9-0693-4b1a-9544-6350fdb70dad",
   "metadata": {},
   "outputs": [],
   "source": [
    "ge_anndata.var[\"ID\"] = ge_anndata.var[\"gene_id\"]\n",
    "ge_anndata.var[\"modality\"] = \"Gene_Expression\"\n",
    "\n",
    "atse_anndata.var[\"ID\"] = atse_anndata.var[\"junction_id\"]\n",
    "atse_anndata.var[\"modality\"] = \"Splicing\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "62c7662c-f480-4749-a376-8e5ca7485986",
   "metadata": {},
   "source": [
    "## 4. Create a Common `.obs` DataFrame\n",
    "\n",
    "You can decide which AnnData’s `.obs` to use (or merge them) if both contain the same information.\n",
    "Here we assume ATSE and GE have matching `obs` indices; we take the ATSE `obs`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60504e70-7f51-4780-82d5-47664296c1e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common obs shape: (157418, 14)\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "common_obs = atse_anndata.obs.copy()\n",
    "common_obs[\"modality\"] = \"paired\"  # if needed; adjust as required\n",
    "print(\"Common obs shape:\", common_obs.shape)\n",
    "\n",
    "# Update both AnnData objects:\n",
    "ge_anndata.obs = common_obs.copy()\n",
    "atse_anndata.obs = common_obs.copy()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "140f8f53-73f3-446a-9454-fe13ed03f875",
   "metadata": {},
   "source": [
    "## 5. Compute or Fix Splicing `junc_ratio` Layer\n",
    "\n",
    "Here we check if `junc_ratio` needs to be recomputed. It is computed as:\n",
    "`junc_ratio = cell_by_junction_matrix / cell_by_cluster_matrix`\n",
    "and any NaNs/Inf values are replaced by zeros.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c48f5b76-9f53-458a-b377-8abffb02f341",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% [markdown]\n",
    "# ### 5.1 Build junc_ratio + psi_mask on the filtered data\n",
    "\n",
    "# %%\n",
    "import numpy as np\n",
    "from scipy import sparse\n",
    "from scipy.sparse import csr_matrix, issparse\n",
    "import gc\n",
    "\n",
    "# grab the splicing modality\n",
    "splicing = atse_anndata  # if you later rename it to 'splicing', otherwise: atse_anndata\n",
    "\n",
    "cell_by_junc    = splicing.layers[\"cell_by_junction_matrix\"]\n",
    "cell_by_cluster = splicing.layers[\"cell_by_cluster_matrix\"]\n",
    "\n",
    "# 1) ensure CSR format\n",
    "if not issparse(cell_by_junc):\n",
    "    cell_by_junc = csr_matrix(cell_by_junc)\n",
    "if not issparse(cell_by_cluster):\n",
    "    cell_by_cluster = csr_matrix(cell_by_cluster)\n",
    "\n",
    "# 2) build psi_mask (1 wherever cluster>0)\n",
    "mask = cell_by_cluster.copy()\n",
    "mask.data = np.ones_like(mask.data, dtype=np.uint8)\n",
    "splicing.layers[\"psi_mask\"] = mask\n",
    "\n",
    "# 3) compute junc_ratio = junction / cluster, nan→0\n",
    "cj = cell_by_junc.toarray()\n",
    "cc = cell_by_cluster.toarray()\n",
    "\n",
    "junc_ratio = np.divide(\n",
    "    cj,\n",
    "    cc,\n",
    "    out=np.zeros_like(cj, dtype=float),\n",
    "    where=(cc != 0),\n",
    ")\n",
    "# 4) assign back as dense or sparse (dense is fine)\n",
    "splicing.layers[\"junc_ratio\"] = junc_ratio\n",
    "\n",
    "print(\"New splicing layers:\", list(splicing.layers.keys()))\n",
    "print(f\"  junc_ratio shape: {junc_ratio.shape}, psi_mask nnz: {mask.nnz}\")\n",
    "\n",
    "# 5) cleanup\n",
    "del cell_by_junc, cell_by_cluster, cj, cc, mask\n",
    "gc.collect()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1acb5f97",
   "metadata": {},
   "source": [
    "## 6. Create a MuData Object\n",
    "\n",
    "Instead of stacking into one AnnData, we create a MuData container.\n",
    "\n",
    "For MULTIVISPLICE, the new setup expects modalities with the following keys:\n",
    "- `rna` : gene expression counts,\n",
    "- `junc_ratio` : raw splicing/junction count data,\n",
    "- `cell_by_junction_matrix` and `cell_by_cluster_matrix` as additional layers.\n",
    "\n",
    "We can use the GE AnnData for gene expression and the ATSE AnnData for all splicing-related data.\n",
    "(If needed, make copies so that modalities are independent.)\n",
    "\n",
    "\n",
    "Option 1: Use the GE AnnData for RNA and the ATSE AnnData for splicing modalities.\n",
    "(You can also combine or pre-process further if desired.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df66d53-f486-48f3-aefe-583561d25080",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/home/svaidyanathan/miniconda3/envs/scvi-env/lib/python3.12/site-packages/mudata/_core/mudata.py:1531: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/gpfs/commons/home/svaidyanathan/miniconda3/envs/scvi-env/lib/python3.12/site-packages/mudata/_core/mudata.py:931: UserWarning: Cannot join columns with the same name because var_names are intersecting.\n",
      "  warnings.warn(\n",
      "/gpfs/commons/home/svaidyanathan/miniconda3/envs/scvi-env/lib/python3.12/site-packages/mudata/_core/mudata.py:1429: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MuData object created with modalities: ['rna', 'splicing']\n"
     ]
    }
   ],
   "source": [
    "mdata = mu.MuData({\n",
    "    \"rna\": ge_anndata,\n",
    "    \"splicing\": atse_anndata\n",
    "})\n",
    "\n",
    "# assert \"library_size\" in ge_anndata.obs, \"'library_size' not found in ge_anndata.obs\"\n",
    "mdata.obsm[\"X_library_size\"] = ge_anndata.obsm[\"X_library_size\"]\n",
    "\n",
    "# # Confirm it's stored correctly\n",
    "# print(\"Library size moved to mdata.obsm['library_size'] with shape:\", mdata.obsm[\"library_size\"].shape)\n",
    "\n",
    "\n",
    "# List of shared obs fields to pull up\n",
    "shared_obs_keys = [\n",
    "    'cell_id', 'age', 'cell_ontology_class', 'mouse.id', 'sex', 'tissue', 'dataset', 'broad_cell_type', 'cell_id_index', 'cell_name', 'modality'\n",
    "]\n",
    "\n",
    "# We'll assume 'rna' modality has them all and they match 'splicing'\n",
    "for key in shared_obs_keys:\n",
    "    assert key in mdata[\"rna\"].obs, f\"{key} not found in 'rna' obs\"\n",
    "    assert key in mdata[\"splicing\"].obs, f\"{key} not found in 'splicing' obs\"\n",
    "    assert (mdata[\"rna\"].obs[key] == mdata[\"splicing\"].obs[key]).all(), f\"{key} values differ between modalities\"\n",
    "    mdata.obs[key] = mdata[\"rna\"].obs[key]\n",
    "    \n",
    "print(\"MuData object created with modalities:\", list(mdata.mod.keys()))\n",
    "print(mdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b693aaf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% [markdown]\n",
    "# ## 8. Stratified train/test split\n",
    "\n",
    "# %%\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# 1) Grab all cell IDs and their labels\n",
    "cells      = mdata.obs_names.to_list()\n",
    "cell_types = mdata.obs[\"broad_cell_type\"].values\n",
    "\n",
    "# 2) Split into train (70%) / test (30%) stratified by broad_cell_type\n",
    "train_cells, test_cells = train_test_split(\n",
    "    cells,\n",
    "    test_size=0.30,\n",
    "    random_state=42,\n",
    "    stratify=cell_types,\n",
    ")\n",
    "\n",
    "# 3) Subset the MuData object\n",
    "mdata_train = mdata[train_cells, :].copy()\n",
    "mdata_test  = mdata[test_cells,  :].copy()\n",
    "\n",
    "# 4) (Optional) If you only need AnnData for the RNA modality:\n",
    "rna_train = mdata_train[\"rna\"]\n",
    "rna_test  = mdata_test[\"rna\"]\n",
    "\n",
    "# 5) Write out both splits\n",
    "mdata_train.write(ROOT_PATH + \"train_70_30_ge_splice_combined_20250513_035938.h5mu\")\n",
    "mdata_test.write( ROOT_PATH + \"test_30_70_ge_splice_combined_20250513_035938.h5mu\")\n",
    "\n",
    "print(\n",
    "    f\"Training cells: {mdata_train.n_obs} ({len(train_cells)})\\n\"\n",
    "    f\"Testing  cells: {mdata_test.n_obs} ({len(test_cells)})\"\n",
    ")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d262328e",
   "metadata": {},
   "source": [
    "## 7. Write Out the Final MuData Object\n",
    "\n",
    "The combined MuData object is now ready for use with `MULTIVISPLICE`. Save it as an H5MU file."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "65276aec",
   "metadata": {},
   "source": [
    "## 8. Verify the Output\n",
    "\n",
    "Read the MuData object back in to ensure everything is correct."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scvi-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
